### PR TITLE
Convert `int constexpr` to `enum` : EconomicLifeCycleCost.*

### PR DESCRIPTION
**Pull request by jmythms**
Pull request overview
---------------------
Convert economic life cycle cost categories from `int constexpr` to `enum`. 

I used plain enums in this PR so that I did not have to use `static_cast`. 

Explaination for Table diffs:
The rows just got rearranged when I rearranged the enums. Rearranged the enums so I could use loops instead of unrolled code for generating the tables.  Screenshot from the ABUPS (.HTML) file generated when using the 5ZoneEconomicsTariffAndLifeCycleCosts.idf file:
image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: I had to use +1 for the third argument in the getObjectItem() function so the vector does not go out of bounds.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:1299]: Enums start at `0` so converted `elcc->CashFlow` from `EPVector` to `std::vector`.

**Comment by amirroth** [comment]
Is it worthwhile making comments on this module since it is tagged for deprecation?

**Comment by mjwitte** [comment]
I'm pretty sure we discussed deprecating ComponentCost:\*, but not LifeCycleCost:\*

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:1299]: If `CashFlow` is really indexed by an `enum` then you can convert it to `std::array`. This applies to all arrays indexed by `enum`.

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.hh]: This thing can be a `constexpr std::array<std::string view, 12>` lookup, it shouldn't be a function. It should also probably be in a central `utility` or `calendar` module.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: `getEnumerationValue`

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: `getEnumerationValue`

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: `getEnumerationValue`

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: Same.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: Same.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: Same.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: This can be done using a table lookup rather than a `switch/case`.

```cpp
constexpr std::array<Real64, DiscConv::Num> DiscConv2EffectiveYearAdjustment = {1.0, 0.5, 0.0};
effectiveYear = double(jYear) - DiscConv2EffectiveYearAdjustment[elcc->discountConvention]; 
```

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:1608]: 👍

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:2046]: Can also be done using a table lookup rather than control flow.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:1745]: 👍 👍

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: Table lookup.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: This can be done with a loop rather than unrolled code.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:2184]: 👍 👍 👍

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [inline src/EnergyPlus/EconomicLifeCycleCost.cc]: Can be done with a loop rather than unrolled code.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.cc]: Thank you!

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.cc]: This is a bit weird since it is not looping through all the CostCategory enums and the index enums are not in the same order. We could probably divide it into three parts, right? 

            tableBody(1, iYear) = RealToStr(elcc->CashFlow[CostCategory::Energy].yrAmount(iYear), 2);
            tableBody(2, iYear) = RealToStr(elcc->CashFlow[CostCategory::Water].yrAmount(iYear), 2);
            for (int i = 0; i < 6; ++i) {
                        tableBody(i+3, iYear) = RealToStr(elcc->CashFlow[i].yrAmount(iYear), 2);
            }
            tableBody(10, iYear) = RealToStr(elcc->CashFlow[CostCategory::TotOper].yrAmount(iYear), 2);

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.cc]: Again, the indexing order is a bit different from the enum, so making this into a loop will cause table diffs. I will leave this to the end to show that this is what is causing the diffs. Or do it in a different PR.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:1299]: Only part of `CashFlow` is indexed by an enum... It is more of a `std::vector`/`Array1D` like container that can [change size](https://github.com/NREL/EnergyPlus/blob/00d968c57b48ff354fdc123d00e782e18842f1ba/src/EnergyPlus/EconomicLifeCycleCost.cc#L1225) based on the number of items in the IDF. 

More details:
[As shown in this line](https://github.com/NREL/EnergyPlus/blob/00d968c57b48ff354fdc123d00e782e18842f1ba/src/EnergyPlus/EconomicLifeCycleCost.cc#L1219), numCashFlow is the size of CashFlow, : 
`elcc->numCashFlow = countOfCostCat + elcc->numRecurringCosts + elcc->numNonrecurringCost + elcc->numResourcesUsed;`. 
In that equation, `countOfCostCat` is always constant and equal to the number of enum values, and that part of `CashFlow` is indexed by the enum. The other terms (`numRecurringCosts`, `numNonrecurringCost`, `numResourcesUsed`) depend on the IDF. 

Should I pull out the `countOfCostCat` part and convert that to `std::array`, and then continue down that path?

Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:1299]: I see, thanks for that clarification. As you were.

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (jmythms) [inline src/EnergyPlus/EconomicLifeCycleCost.hh:260]: Is there a better way to do this?

Comment (Myoldmopar) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: Maybe, but I'm not sure if it is worth it or not.  C++ has a `toupper` function that operates on individual characters, but not on a string.  So you'd have to iterate over the characters, and you'd need to construct a string while doing it.  Maybe @lefticus has a fancy idea for how to do this most efficiently, but unless this is a widespread issue on this PR, I'd just put a pin in it.  Getting it into `constexpr string_view` is already such an improvement over the past...

Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: This is the fastest and cheapest way to do this unless the `MakeUpper` function can itself be made `constexpr`. I don't think it can, but it may be worth a quick test. Even then the implementation wouldn't be cheaper, it would have the same cost but maybe the code would be a bit clearer. Either way, it's not worth holding up this PR.

Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: Thank you for the inputs and insights. 
No, it's not a widespread issue, just felt that it felt a bit weird pulling these out as individual variables. 

Will investigate changing `MakeUpper` into a `constexpr` function separately.

Comment (Myoldmopar) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: https://godbolt.org/z/nKcEf5eev  I've given it 10 minutes and haven't figured out a path forward on it.  I assume that means it is not possible.  :laughing: :rofl: JK, but for real that's a starting point if anyone wants to try to make a clean solution.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:396]: @mjwitte maybe you can remind me here...can we not count on the input being all caps at this point?  Assuming we stick with a case-insensitive input format, even for epJSON values, we should just make sure the input processor upper cases everything, even the epJSON blob.  (Of course not the fields with `\retaincase`).  

It would be really nice to just know for sure that the AlphaArray was already upper case then we could eliminate so many MakeUPPERCase function calls and clean up so many chunks of code.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:673]: All these changes seem fine really.  I definitely look forward to the BITF stuff becoming constexpr functions instead of macros.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:298]: Nice!  So much cleaner, and another function gone.

File: tst/EnergyPlus/unit/EconomicLifeCycleCost.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/EconomicLifeCycleCost.unit.cc:503]: I probably would have just removed this unit test entirely.  But it's fine, it's extra checking of the getEnumerationValue function.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:2171]: No need to put parentheses around the case labels.  No need to hold it up, but in the future don't add them.

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: Where did you learn how to program?

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.hh:117]: This isn't the easiest thing to read/maintain, but I think it's a net gain.  The huge IF block, non-constexpr, assignments weren't great either.  Maybe there is a better way to format the tabular data ... I dunno.  No need to address it now.

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:260]: More seriously, `makeUpperCase` cannot be made `constexpr` because it has to allocate memory for the target string so this path is closed. Just leave as is for now.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EconomicLifeCycleCost.cc:2250]: I see, so this is why the table diffs appear.  I think that is acceptable and these loops are much nicer than the manually listed out chunks of code.

**Review by Myoldmopar — APPROVED** [review]
This is really great.  I don't see any red flags here, this is just tons of really great cleanup.  I will do some final testing locally but as far as I can tell this is good to go in.  If anyone else wants any specific tests/demos/changes, speak up soon.

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (mjwitte) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:396]: At present, there's no global pass to push the input strings to UPPERcase. `getObjectItem` and `getAlphaFieldValue` return UPPERcase strings. Direct access of the JSON data does not.

File: tst/EnergyPlus/unit/EconomicLifeCycleCost.unit.cc
Comment (jmythms) [reply tst/EnergyPlus/unit/EconomicLifeCycleCost.unit.cc:503]: Got it, will remember next time!

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:396]: I don't think that is strictly necessary. There is nothing wrong with `makeUpperCase` as long as you only do it once per string. The most important thing is to avoid case-insensitive comparisons and there are other ways to do that. See [this issue](https://github.com/NREL/EnergyPlus/issues/9183).

Comment (Myoldmopar) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:396]: That's what I was afraid of, but I also see a good possibility there.  If the IP just makes one quick pass to upper case all the (non-retain-case) fields in the IDF or epJSON, then we would get rid of a huge amount of these `MakeUPPERCase` calls spread out all over the code.  And also it would make it perfectly clear to the developers that the data **will** be upper case as soon as the input processor is done with it.  No ambiguity that leads to extraneous string operations during input processing because the dev isn't sure if it will be upper case or not.  

I think this would be a great discussion to have, not here, but soon.  Thanks!

Comment (Myoldmopar) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:396]: Thanks @amirroth, I think I am in full agreement that there is nothing wrong with MakeUPPERCase, as long as it is only done once, but the problem is that we aren't quite sure if it's been done or not, so it ends up being done extra times.  If the input processor just made it upper case one time then we could be absolutely sure it's upper case and eliminate many MakeUPPERCase calls spread out of the code.  Anyway, nothing for this PR.  Thanks!

Comment (amirroth) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:396]: I agree. If `MakeUpperCase` is done in input processor by default that's even better, but what I was saying is that we can eliminate a lot of the expensive case-insensitive comparisons even with no changes to the input processor or JSON module, because the problem is not `MakeUpperCase` inherently, it's the fact that SameString essentially does a `MakeUpperCase` on both inputs and it is called by many search routines.

**Comment by Myoldmopar** [comment]
Merging in 10 minutes unless someone speaks up otherwise.

File: src/EnergyPlus/EconomicLifeCycleCost.hh
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.hh:117]: Would it be better to split stuff up to improve readability? Something like,
```
constexpr std::array<string_view> smallTable1;
constexpr std::array<string_view> smallTable2;

constexpr std::array<std:array> bigTable {smallTable1, smallTable2};
```

File: src/EnergyPlus/EconomicLifeCycleCost.cc
Comment (jmythms) [reply src/EnergyPlus/EconomicLifeCycleCost.cc:2171]: Got it. Happened because I memorized this pattern. Will not add them in the future. 
image

**Comment by jmythms** [comment]
Thanks so much!

**Pull request merged**
